pkg/kv/mem/store.go

```diff
@@ -0,0 +1,169 @@
+package mem
+
+import (
+	"bytes"
+	"context"
+	"sort"
+	"sync"
+
+	"github.com/treeverse/lakefs/pkg/kv"
+)
+
+type Driver struct{}
+
+type Store struct {
+	m    map[string][]byte
+	keys []string
+	mu   sync.RWMutex
+}
+
+type Entries struct {
+	entry *kv.Entry
+	err   error
+	start []byte
+	first bool
+	store *Store
+}
+
+const DriverName = "mem"
+
+//nolint:gochecknoinits
+func init() {
+	kv.Register(DriverName, &Driver{})
+}
+
+func (d *Driver) Open(_ context.Context, _ string) (kv.Store, error) {
+	return &Store{
+		m:    make(map[string][]byte),
+		keys: []string{},
+	}, nil
+}
+
+func (s *Store) Get(_ context.Context, key []byte) ([]byte, error) {
+	if key == nil {
+		return nil, kv.ErrMissingKey
+	}
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	value, ok := s.m[string(key)]
+	if !ok {
+		return nil, kv.ErrNotFound
+	}
+	return value, nil
+}
+
+func (s *Store) Set(_ context.Context, key, value []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if _, found := s.m[string(key)]; !found {
+		s.insertNewKey(key)
+	}
+	s.m[string(key)] = value
+	return nil
+}
+
+// insertNewKey insert new key into keys - insert into sorted slice
+func (s *Store) insertNewKey(key []byte) {
+	idx := sort.SearchStrings(s.keys, string(key))
+	if idx == len(s.keys) {
+		s.keys = append(s.keys, string(key))
+	} else {
+		s.keys = append(s.keys, "")
+		copy(s.keys[idx+1:], s.keys[idx:])
+		s.keys[idx] = string(key)
+	}
+}
+
+func (s *Store) SetIf(_ context.Context, key, value, valuePredicate []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	curr := s.m[string(key)]
+	if valuePredicate == nil {
+		if curr != nil {
+			return kv.ErrPredicateFailed
+		}
+		s.insertNewKey(key)
+	} else if !bytes.Equal(valuePredicate, curr) {
+		return kv.ErrPredicateFailed
+	}
+	s.m[string(key)] = value
+	return nil
+}
+
+func (s *Store) Delete(_ context.Context, key []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if _, found := s.m[string(key)]; !found {
+		return nil
+	}
+	idx := sort.SearchStrings(s.keys, string(key))
+	if idx < len(s.keys) && s.keys[idx] == string(key) {
+		s.keys = append(s.keys[:idx], s.keys[idx+1:]...)
+	}
+	delete(s.m, string(key))
+	return nil
+}
+
+func (s *Store) Scan(_ context.Context, start []byte) (kv.Entries, error) {
+	return &Entries{
+		store: s,
+		first: true,
+		start: start,
+	}, nil
+}
+
+func (s *Store) Close() {}
+
+func (e *Entries) Next() bool {
+	if e.start == nil {
+		return false
+	}
```

Comment:
Should we not check for e.err as well? For example I can perform Close() and then Next() and it will work
Same for Entry()?
Same question for postgres implementation

Reply:
1. true - fixed in the 4min ago commit
2. entry should be covered as Next implementation will set the right value. There is no garenty on using any object after calling Close. I'll add a code doc to exlpain that.
3. same handling can be found in postgres too

Reply:
@N-o-Z found a case where entry is not set to nil - fixing it and add test code to verify

Reply:
ok, false alarm - it was a case where we break the iteration because we got to an entry which doesn't match the same prefix.
Think if a helper method on the iterator should help write code that using it.